ouroboros-network-framework/src/Ouroboros/Network/Util/TaggedObserve.hs

```diff
@@ -0,0 +1,37 @@
+{-# LANGUAGE LambdaCase #-}
+module Ouroboros.Network.Util.TaggedObserve (
+      TaggedObservable (..)
+    , matchTaggedObservations
+
+      -- reexports
+    , ObserveIndicator (..)
+    , Observable (..)
+    , matchObservations
+    ) where
+
+import           Control.Tracer
+import           Control.Tracer.Observe
+
+-- Like Observable but with a t tag type field too.
+data TaggedObservable t s e d = TOStart t s
+                              | TOEnd t e (Maybe d)
+                              --           ^^ holds the difference between start and end
+matchTaggedObservations
+    :: Monad m
+    => (t -> m (Maybe s))
+    -> (t -> s -> m ())
+    -> (s -> e -> d)
+    -> Tracer m (TaggedObservable t s e d)
+    -> Tracer m (TaggedObservable t s e d)
+matchTaggedObservations getStart putStart f tr = Tracer $ \case
+    obs@(TOStart t s) -> do
+        putStart t s
+        traceWith tr obs
+    (TOEnd t e _) -> do
+        before <- getStart t
+        traceWith tr $ TOEnd t e $ fmap (flip f e) before
```

Comment:
What I like about this implementation is that it has the property that if we forget to put a matching `TOStart` or `TOEnd` the next computation will return `Nothing` (if I understand it correctly). Which is a nice property to have.

Reply:
That behavior is the same as the untagged implementation.